chips/earlgrey/src/registers/top_earlgrey.rs

```diff
@@ -2,10 +2,10 @@
 // SPDX-License-Identifier: Apache-2.0 OR MIT
 // Copyright lowRISC contributors 2023.
 
-// Built for Earlgrey-M2.5.1-RC1-389-g21ce4e9761
-// https://github.com/lowRISC/opentitan/tree/21ce4e9761abdf5c919b46e5ae64a5a8e24992f7
+// Built for Earlgrey-M2.5.2-RC0-904-g9a5280ee33
+// https://github.com/lowRISC/opentitan/tree/9a5280ee3331bc20f2659eee37a56f270454bdcf
```

Comment:
We should also update the [Makefile sha](https://github.com/tock/tock/blob/3df5b00e00121f645740ec39441f93cb101feec1/boards/opentitan/earlgrey-cw310/Makefile#L19). It looks like there's a mismatch already too.

Reply:
I think it's ok if those SHAs don't match. Since the HW isn't changing, we shouldn't need to update these definitions for correctness, but changes like this renaming might come up time-to-time.

Reply:
This commit is based on `upstream/master OpenTitan` version of file  `top_earlgrey.rs`.
If it will be possible i would like to stay on top of Earlgrey ES branch because this branch will be used for HW samples.  Changes in `top_earlgrey.rs`  are pure syntax sugar.

Reply:
As long as this file (top_earlgrey.rs) came from the `earlgrey_es` branch, then we don't need to update the SHA referenced in the Makefile.  There are no hardware changes to anything on the `earlgrey_es` branch (apart from the already-known FPGA clocking changes).

The SHA in the Makefile (`7df72d61...`) refers to the correct hardware version and the bitstream one should download from https://storage.googleapis.com/opentitan-bitstreams/

Reply:
Why not just update them both at the same time? If there are no HW changes it's a very simple change in Tock and that way everything is clear and in sync.

It seems strange to be using different hardware revisions for the registers and the bitstreams

Reply:
The bitstream should come from the `earlgrey_es` branch, as that branch is being used for hardware production.

The register definition generation tool is not yet stable. We are still refining it (as you can see in this PR), and plan to continue to work on it. For that reason, the registers should be generated using the `master` branch.

Which means the exact commits won't match, but that is fine.

Reply:
Why not just apply the change to the same place and update all of the SHAs here? What happens when a hardware change goes into `master` and then gets propagated here but then doesn't match what we are flashing to the board?

Reply:
I agree with @jrvanwhy - we should keep bitstream SHA on  `earlgrey_es`  and update software related things to upstream `master` branch if they don't collide with hardware changes.
Hardware is more that end user interface (register layout), using upstream version of bitstream can cause some trouble in debugging issue, because you basically debug different implementation of hardware.

However, due to the imperfections of the data generation process for Rust, it is best to use a upstream version. The version with `earlgrey_es` does not yet implement the Copy,Clone, PartialEq,Eq traits also we have repetitions of chip name prefix in constans, with can cause problem with newer chip version.

Reply:
Generating registers is separate from the HW bitstream, and I don't see any issue with the commits being different. That actually makes sense to me, if the HW is frozen then the tools shouldn't need to be fixed to the same commit. Much in the same way that libtock-c and the kernel evolve separately.

Reply:
> Why not just apply the change to the same place and update all of the SHAs here? What happens when a hardware change goes into `master` and then gets propagated here but then doesn't match what we are flashing to the board?

Cherrypicking software changes into `earlgrey_es` requires extra effort, and making a habit of it would prevent the `earlgrey_es` branch from stabilizing.

In the long run, we may need other solutions for generating Earlgrey registers (such as by running a latest-`master` register generation tool on register definitions from the `earlgrey_es` branch), but we're not there yet and using `master` to generate register definitions is what makes sense now.

Reply:
> Why not just update them both at the same time? If there are no HW changes it's a very simple change in Tock and that way everything is clear and in sync.
>
>It seems strange to be using different hardware revisions for the registers and the bitstreams

The have been some software and infrastructure improvements on `master` that make cherry-picking the reggen/topgen updates onto `earlgrey_es` impractical.  Since HW changes are nearly forbidden[^1] on `master`, we're going to generate these constants from master.

We want to keep the SHA in the makefile the same because that SHA refers to the version of the codebase that would construct the exact bitstream, verilator model, test_rom and mask_rom that are being used for the chip tape-out.  If you're running tock on the FPGA or in a simulator, **that** is the version you want.

[^1] We still permit ECOs on `master`.  The ECOs are typically done on `earlgrey_es` in a way that is compatible with the floor-plan and physical layout of the chip.  Those ECOs are then re-done on `master` the "right" way (rather than bound by the tape-out constraints in `earlgrey_es`).

Reply:
> Generating registers is separate from the HW bitstream, and I don't see any issue with the commits being different.

How is it separate? The whole idea is that the registers are generated from the RTL so they will update together. By keeping them in sync we know the registers match the hardware and it's all automatic

> Since HW changes are nearly forbidden[^1] on `master`, we're going to generate these constants from master.

This doesn't really explain why they can't be in sync. If a hardware change happens on master the register definition will change, but the bitstream we are using won't? That seems prone to breakage.

Like you say, there isn't going to be a lot of churn on master, so why not just update it?

> That actually makes sense to me, if the HW is frozen then the tools shouldn't need to be fixed to the same commit. Much in the same way that libtock-c and the kernel evolve separately.

But the hardware isn't frozen. There has already been one change  since tapeout for FPGAs and @cfrantz  just said that some changes can go in.

libtock-c and the kernel are both written to be API compatible. A change in the hardware would generally be an "ABI breakage", so it doesn't seem exactly the same

> We want to keep the SHA in the makefile the same because that SHA refers to the version of the codebase that would construct the exact bitstream, verilator model, test_rom and mask_rom that are being used for the chip tape-out. If you're running tock on the FPGA or in a simulator, **that** is the version you want.

Isn't that not the case? We have already bumped the SHA we use in the Makefile since tapeout to get faster clock frequencies. Why can't we do the same to update register layouts.

We are already bumping the Makefile from tape out: https://github.com/tock/tock/pull/3703/files. So we are no longer targeting the bitstream from hardware production. It seems strange that it wasn't an issue for https://github.com/tock/tock/pull/3703 but it's an issue now. What am I missing here?

Reply:
@alistair23 I don't know how much do you know about OpenTitan development process, i see this topic is hot and opinionated - we should focus more on possible solutions. 

From my point of view if this PR will be discarded current approach (`master`) will be needed https://github.com/lowRISC/opentitan/pull/19988

Other option is to rewrite code in such way that all traits will be implemented manually.

Reply:
> > Generating registers is separate from the HW bitstream, and I don't see any issue with the commits being different.
> 
> How is it separate? The whole idea is that the registers are generated from the RTL so they will update together. By keeping them in sync we know the registers match the hardware and it's all automatic

If I'm understanding correctly, what is changing in new commits isn't the actual register meanings, but instead the tool. Updating the tool for the auto code generation shouldn't require new hardware testing.

Reply:
> > Generating registers is separate from the HW bitstream, and I don't see any issue with the commits being different.
> 
> How is it separate? The whole idea is that the registers are generated from the RTL so they will update together. By keeping them in sync we know the registers match the hardware and it's all automatic

These changes to the register definition are *software* changes, not *hardware* changes. The tool that processes the register description files (which are stable) has proved to generate better definitions. We want to use the better definitions here in Tock, so we want to generate the definitions using the tool from latest `master`.

> > Since HW changes are nearly forbidden[^1] on `master`, we're going to generate these constants from master.
> 
> This doesn't really explain why they can't be in sync. If a hardware change happens on master the register definition will change, but the bitstream we are using won't? That seems prone to breakage.

If hardware changes were to happen on master, sure, but (except for ECOs) hardware changes are banned on `master` too.

> Like you say, there isn't going to be a lot of churn on master, so why not just update it?

We could pull the bitstream from `master`, but our goal here is to run Tock on the real hardware, and `earlgrey_es` is the closest bitstream we have to that.

> > That actually makes sense to me, if the HW is frozen then the tools shouldn't need to be fixed to the same commit. Much in the same way that libtock-c and the kernel evolve separately.
> 
> But the hardware isn't frozen. There has already been one change since tapeout for FPGAs and @cfrantz just said that some changes can go in.

ECOs are very unlikely to result in register changes.

> libtock-c and the kernel are both written to be API compatible. A change in the hardware would generally be an "ABI breakage", so it doesn't seem exactly the same
> 
> > We want to keep the SHA in the makefile the same because that SHA refers to the version of the codebase that would construct the exact bitstream, verilator model, test_rom and mask_rom that are being used for the chip tape-out. If you're running tock on the FPGA or in a simulator, **that** is the version you want.
> 
> Isn't that not the case? We have already bumped the SHA we use in the Makefile since tapeout to get faster clock frequencies. Why can't we do the same to update register layouts.
> 
> We are already bumping the Makefile from tape out: https://github.com/tock/tock/pull/3703/files. So we are no longer targeting the bitstream from hardware production. It seems strange that it wasn't an issue for #3703 but it's an issue now. What am I missing here?

The SHA is #3703 is just a newer commit from the `earlgrey_es` branch. You're asking us to either copy a bunch of software changes into `earlgrey_es` or use a hardware bitstream from `master`, and both of those ideas have issues with no upside other than "the SHAs match".

Reply:
> > > Generating registers is separate from the HW bitstream, and I don't see any issue with the commits being different.
> > 
> > 
> > How is it separate? The whole idea is that the registers are generated from the RTL so they will update together. By keeping them in sync we know the registers match the hardware and it's all automatic
> 
> If I'm understanding correctly, what is changing in new commits isn't the actual register meanings, but instead the tool. Updating the tool for the auto code generation shouldn't require new hardware testing.

Yes, exactly.

Reply:
I understand that the `master` OT branch has software changes for the register layout generation.

Like Johnathon just said, there might be hardware changes on master to fix issues. When/If that happens and we bump the register layout again the hardware used to generate the registers and the actual bitstream will be out of sync. I understand you expect it to be unlikey that there will be register changes, but we don't know for sure that won't happen.

Hardware wise `earlgrey_es` and `master` should be practically the same as you just said. There are no new features being developed on master, so we don't expect a lot of churn.

So what is wrong with just updating the bitstream SHA to point to master? You are worried it is out of sync? But then the registers generated would be out of sync.

> The SHA is #3703 is just a newer commit from the `earlgrey_es` branch. You're asking us to either copy a bunch of software changes into `earlgrey_es` or use a hardware bitstream from `master`, and both of those ideas have issues with no upside other than "the SHAs match".

I never said anything about copying. I just think that the register definitions we are using should match the hardware we are targeting. Which means they should be generated from the same RTL source. That way we know they are in sync.

We wouldn't generate Rustdocs for a release from a different source for example, I don't see why this is any different.

Reply:
The PR with backport SW  changes to `earlgrey_es`was merged:
https://github.com/lowRISC/opentitan/pull/19988

Updated:
 - [x] makefile
 -  [x] register
 -  [x] soc constans
 - [ ] README
 
 let me know if i miss somthing